mslib/msui/_tests/test_mscolab.py

```diff
@@ -211,16 +247,44 @@ def test_set_exported_file(self, mockopen, mockmessage):
         QtWidgets.QApplication.processEvents()
         assert self.window.listProjects.model().rowCount() == 1
 
-    def test_add_project(self):
+    @mock.patch("PyQt5.QtWidgets.QErrorMessage")
+    def test_add_project(self, mockbox):
         # ToDo test needs to be independent from test_user_delete
         self._connect_to_mscolab()
         self._create_user("something", "something@something.org", "something")
         self._login("something@something.org", "something")
         assert self.window.label.text() == 'Welcome, something'
         assert self.window.loginWidget.isVisible() is False
         self._create_project("Alpha", "Description Alpha")
+        assert mockbox.return_value.showMessage.call_count == 2
+        with mock.patch("PyQt5.QtWidgets.QLineEdit.text", return_value=None):
+            self._create_project("Alpha2", "Description Alpha")
+        with mock.patch("PyQt5.QtWidgets.QTextEdit.toPlainText", return_value=None):
+            self._create_project("Alpha3", "Description Alpha")
+        self._create_project("/", "Description Alpha")
+        assert mockbox.return_value.showMessage.call_count == 5
         assert self.window.listProjects.model().rowCount() == 1
 
+    @mock.patch("mslib.msui.mscolab.MSCOLAB_AuthenticationDialog.exec_", return_value=QtWidgets.QDialog.Accepted)
+    @mock.patch("PyQt5.QtWidgets.QErrorMessage")
+    def test_failed_authorize(self, mockbox, mockauth):
+        class response:
+            def __init__(self, code, text):
+                self.status_code = code
+                self.text = text
+
+        self._connect_to_mscolab()
+        with mock.patch("requests.Session.post", new=ExceptionMock(requests.exceptions.ConnectionError).raise_exc):
+            self._login()
+        with mock.patch("requests.Session.post", return_value=response(201, "False")):
+            self._login()
+        with mock.patch("requests.Session.post", return_value=response(401, "False")):
+            self._login()
+
+        # No return after self.error_dialog.showMessage('Oh no, server authentication were incorrect.')
+        # causes 4 instead of 3 messages, I am not sure if this is on purpose.
+        assert mockbox.return_value.showMessage.call_count == 4
```

Comment:
from the naming I would have expected mockauth here, or?

Reply:
QErrorMessage is the one on which showMessage is called, as such that needs to be checked on the mockbox.

Reply:
somehow I thought before that the order of the input arguments is given by the order of the patch commands. Or that was my private rule, not found something in the documentation on a quick search.
The second one is @mock.patch("PyQt5.QtWidgets.QErrorMessage") I automatically would have looked on mockauth.

Reply:
found something, it is by order.
https://stackoverflow.com/questions/33353855/how-does-mock-patch-know-which-parameter-to-use-for-each-mock-object

Reply:
it is correct from, the variable, but for reading I would have epected mockauth connected to the auth dialog mock patch.

Reply:
I wasn't sure about the order either. Basically I found the top decorator is the last function argument.
QErrorMessage is the mockbox and the Auth dialogue is the mockauth.